### DISCUSSION TITLE
Using `apply` when using shims, but without JS

### DISCUSSION THREAD

**Discussion by srid** [tw-in-js/twind#163] [Q&A] [answered]
I have a static site project that doesn't really use JS (except where strictly necessary), and I'm using the twind shim during development because it is a no fuss way to get tailwind working (during production static generation I might use `windicss` to produce a small .css file and include that).

The shim works on element classes, however if I want to create an unique style like the following, how can I make it work when using twind/shim?

```html
<style>
.btn {
  @apply bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50;
}
</style>
```

twind/shim currently ignores the style tags.

**Comment by sastan** [comment, marked as answer]
You can define a plugin for this. There are two ways for static site:

1. Using `<script type="twind-config">`

```html
<script type="module" src="https://cdn.skypack.dev/twind/shim"></script>
<script type="twind-config">
  {
    "plugins": {
      "btn": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
    }
  }
</script>

<button class="btn">Click me</button>
```

2. Using setup of twind/shim

```html
<script type="module">
  import { setup } from "https://cdn.skypack.dev/twind/shim"
  
  setup({
    plugins: {
      btn: "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
    }
  })
</script>

<button class="btn">Click me</button>

```

**Reply by srid** [reply]
This partially addresses the problem, but won't work in the static site generation case (when I wouldn't want to use twind/shim due to [slowness](https://github.com/tw-in-js/twind/discussions/161) anyway).

**Reply by srid** [reply]
On the other hand, if twind had a CLI to generate the final .css like `windicss` does, that would solve my problem. 

All that said, simply supporting "<style>" tag would be much favourable in regards to convention-over-configuration philosophy.

**Reply by danielweck** [reply]
> when I wouldn't want to use twind/shim due to [slowness](https://github.com/tw-in-js/twind/discussions/161) anyway

Well, as demonstrated further down in the discussion thread, the Twind runtime is pretty fast when used appropriately :)

That being said, Twind's creators have indeed stated that the shim is suboptimal, and that static build-time stylesheet generation is preferable in some cases.

I think as things evolve we will see better-documented options on how to use Twind effectively in different integration scenarios.
(the project is still young)

**Reply by sastan** [reply]
> On the other hand, if twind had a CLI to generate the final .css like windicss does, that would solve my problem.

@srid You could try [@twind/cli](https://github.com/tw-in-js/twind-cli)

Create a `twind.config.js` 

```js
export default {
  plugins: {
    btn: "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
  }
}
```

run `twind` and it generates a static css file

**Reply by srid** [reply]
Can you style an entire element using that `plugins` JSON syntax?  For eg., normally I would do this in the style tag as:

```html
<style>
a { 
  @apply purple-500 font-bold hover:underline;
}
</style>
```

Since I'd be dealing with _generated_ HTML (eg: from Markdown library), this capability would be important for me, and it seems you just can't do it the JS way.

**Reply by srid** [reply]
Looks like this is possible as a _global_ style? https://twind.dev/docs/handbook/advanced/plugins.html#inject-global-styles

```js
setup({
  plugins: {
    link: {
      ':global': {
        a: {
          /* global styles for anchors */
        },
      },
      /* element styles */
    },
  },
})
```

**Reply by danielweck** [reply]
If I am not mistaken, the plugin approach requires explicitly referring to the 'link' class name, in your example.
Personally I use the css() function to set global styles.

**Reply by sastan** [reply]
You can add global styles via [preflight](https://twind.dev/docs/handbook/advanced/setup.html#preflight)

```html
<script type="twind-config">
  {
    "preflight": {
       "a": {
          "@apply": "text-blue-500 hover:text-blue-700",
          "css-property": "css-value",
          "&:hover: {},
       } 
    },
    "plugins": {
      "btn": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
    }
  }
</script>
```

**Reply by srid** [reply]
@sastan Thanks! But how do you add this when using shims? 

In another site, I wanted to use twind shim **whilst disabling the preflight**. I put this in head, but it still loads the tailwind preflight,

```html
<script type="twind-config">
    {
      "preflight": false,
      "plugins": {
        "otherClass": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
      }
    }
</script>
<script type="text/javascript" crossorigin="anonymous" src="https://unpkg.com/twind@0.16.9/twind.umd.js"> </script>
<script type="text/javascript" crossorigin="anonymous"
    src="https://unpkg.com/twind@0.16.9/observe/observe.umd.js"> </script>
<script type="text/javascript">
    twindObserve.observe(document.documentElement)
</script>
```

(Not using official shim include, due to the optimized version talked about [here](https://github.com/tw-in-js/twind/discussions/161))

**Reply by danielweck** [reply]
The UMD JavaScript bundles are loaded synchronously, so you should be able to call setup() immediately after the main Twind JavaScript (just before the observe call).

**Reply by danielweck** [reply]
twindSetup(...) should be a global property on the window object (I haven't tried myself, just guessing)

**Reply by sastan** [reply]
The `<script type="twind-config">` is only used by twind/shim (for now).

Please use something like this for now:

```html
<script type="text/javascript" crossorigin="anonymous" src="https://unpkg.com/twind@0.16.9/twind.umd.js"> </script>
<script type="text/javascript" crossorigin="anonymous"
    src="https://unpkg.com/twind@0.16.9/observe/observe.umd.js"> </script>
<script type="text/javascript">
    twind.setup({
      "preflight": false,
      "plugins": {
        "otherClass": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
      }
    })

    twindObserve.observe(document.documentElement)
</script>
```

Or for the shim like experience:

```html
<script type="twind-config">
    {
      "preflight": false,
      "plugins": {
        "otherClass": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
      }
    }
</script>
<script type="text/javascript" crossorigin="anonymous" src="https://unpkg.com/twind@0.16.9/twind.umd.js"> </script>
<script type="text/javascript" crossorigin="anonymous"
    src="https://unpkg.com/twind@0.16.9/observe/observe.umd.js"> </script>
<script type="text/javascript">
    twind.setup(JSON.parse(document.querySelector('script[type="twind-config"]').innerHTML))

    twindObserve.observe(document.documentElement)
</script>
```

**Reply by srid** [reply]
Appreciate the answers! 

Now I can use twind in my static sites, via neuron: https://neuron.zettel.page/tailwind

**Reply by sastan** [reply]
@srid I have added twind to jsdelivr for faster CDN and bundling several modules into one script:

```html
<script crossorigin="anonymous" src="https://cdn.jsdelivr.net/combine/npm/twind/twind.umd.min.js,npm/twind/observe/observe.umd.min.js"></script>
<script>
  twind.setup(
    {
      mode: 'silent',  // Behave well with semantic UI classes
      preflight: false,  // Don't reset semantic UI!
    }
  );
  twindObserve.observe(document.documentElement)
</script>
```

**Reply by danielweck** [reply]
>  jsdelivr for faster CDN and bundling several modules into one script

Good idea.

Note to self (I always forget this): Jason Miller's ECMAScript Module URL bundler does this too:
https://npm.reversehttp.com/twind,twind/shim
(Source code: https://glitch.com/edit/#!/2010?path=app.js%3A1%3A0 )

**Reply by srid** [reply]
@sastan Will that be added to the official docs? Is there an issue to track it, so that I'll know when to update the snippet in https://neuron.zettel.page/tailwind ?

**Reply by sastan** [reply]
It will be in the docs. I hope to get the new docs published this week. This would then simplifying the  MR process for docs changes.

**Reply by srid** [reply]
@sastan How would you configure preflight like this for elements (eg: style all `<a>` elements, regardless of the class)? Currently preflight accepts only class selectors:

```json
      "plugins": {
        "otherClass": "bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50"
      }
```

(still assuming we are disabling the global reset; as I would be using it on top of semantic UI)

**Reply by srid** [reply]
Also, this "otherClass" configuration is more of a hack (when treating it as anything but a plugin), as twind completely replaces "otherClass" with tailwind's classes - when "otherClass" should stay in place. 

So I guess the question is: how do we style elements using any CSS selector by specifying a list of tailwind classes to apply? Normally with tailwind you can do this in `<style>` tags using `@apply`, but twind has no support for it.

**Reply by lukejacksonn** [reply]
Hi @srid if you want to target and style arbitrary classes then Twind does support that, it would look something like this is p/react.

```js

const customStyles = css`
  .otherClass {
    @apply bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50;
  }
`
const Component = () =>
  <div className={customStyles}>
     <button className="otherClass">I'm targeted by a CSS selector with @apply</button>
  </div>
```

**Reply by srid** [reply]
Hey @lukejacksonn - note that I'm using this on a statically generated site. i.e., no JavaScript frameworks; just HTML. I actually use twind to style the pink box in https://notes.srid.ca/now (because the Markdown adds a new element with a new class name) - but I cannot use twind to style *existing* classes used by the static site generator (neuron, in this case).

**Reply by sastan** [reply]
@sid You can do global styling using the preflight or `@global` within a tw call

1. preflight

```html
<script crossorigin="anonymous" src="https://cdn.jsdelivr.net/combine/npm/twind/twind.umd.min.js,npm/twind/observe/observe.umd.min.js"></script>
<script>
  twind.setup(
    {
      mode: 'silent',  // Behave well with semantic UI classes
      // Use a function to disable preflight merging
      preflight: () => {
        return {
           // twind/css would work here as well
           ".otherClass": twind.apply("bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50")
        }
      }
    }
  );
  twindObserve.observe(document.documentElement)
</script>
```

2. `@global` within a tw call

```html
<script crossorigin="anonymous" src="https://cdn.jsdelivr.net/combine/npm/twind/twind.umd.min.js,npm/twind/observe/observe.umd.min.js"></script>
<script>
  twind.setup(
    {
      mode: 'silent',  // Behave well with semantic UI classes
      preflight: false,  // Don't reset semantic UI!
    }
  );

  // twind/css would work here as well
  twind.tw(() => {
    '@global': {
       ".otherClass": twind.apply("bg-purple-200 p-3 border-2 hover:border-4 hover:opacity-50")
    }
  })
  twindObserve.observe(document.documentElement)
</script>
```

**Reply by srid** [reply]
That works for applying those classes, however when I used the following:

```
  twind.setup(
    {
      mode: 'silent',  // Behave well with semantic UI classes
      preflight: () => {
        return {
        }
      }
    }
  );
  twindObserve.observe(document.documentElement)
```

It did reset some of my existing styles. For eg., the black left border line on the post listings in https://notes.srid.ca/ vanished entirely, unless I comment out the `twindObserve.observe` line. 

If I comment out the `preflight` field entirely, then the tailwind reset is fully applied. Even if I use `preflight: false` the border remains vanished. I'm convinced that there is bug somewhere in twind.

**Reply by srid** [reply]
Oh, actually. The conflict arises only for CSS styling I defined in the `<head>` just before twind, and not for styles loaded from URLs. So that's not so bad.

**Reply by srid** [reply]
Wait, but how would you specify `!important` to these applied classes?

**Reply by sastan** [reply]
`apply` has some issues currently (like `!` does not work). I'm working on that. In the meantime you can use plain CSS:

```html
<script crossorigin="anonymous" src="https://cdn.jsdelivr.net/combine/npm/twind/twind.umd.min.js,npm/twind/observe/observe.umd.min.js"></script>
<script>
  twind.setup(
    {
      mode: 'silent',  // Behave well with semantic UI classes
      // Use a function to disable preflight merging
      preflight: () => {
        return {
           // twind/css would work here as well
           ".otherClass": {
             ‘@apply': '....',
             color: 'red !important',
           }
        }
      }
    }
  );
  twindObserve.observe(document.documentElement)
</script>
```

**Reply by srid** [reply]
I found a workaround. Using more specific selectors (CSS specificity) obviated the need for !important.

Voila, https://tribe.srid.ca/@srid/posts/A62qdzCRqf3tdKsONU

> In the meantime you can use plain CSS

Didn't know that was possible, good to know!
